config/kernel.m4

```diff
@@ -772,11 +772,20 @@ dnl # $2 - source
 dnl # $3 - run on success (valid .ko generated)
 dnl # $4 - run on failure (unable to compile)
 dnl #
+dnl # When configuring as builtin (--enable-linux-builtin) for kernels
+dnl # without loadable module support (CONFIG_MODULES=n) only the object
+dnl # file is created.  See ZFS_LINUX_TEST_COMPILE_ALL for details.
+dnl #
 AC_DEFUN([ZFS_LINUX_TRY_COMPILE], [
-	ZFS_LINUX_COMPILE_IFELSE(
-	    [ZFS_LINUX_TEST_PROGRAM([[$1]], [[$2]])],
-	    [test -f build/conftest/conftest.ko],
-	    [$3], [$4])
+	AS_IF([test "x$enable_linux_builtin" = "xyes"], [
+		ZFS_LINUX_COMPILE_IFELSE(
+		    [ZFS_LINUX_TEST_PROGRAM([[$1]], [[$2]])],
+		    [test -f build/conftest/conftest.o], [$3], [$4])
+	], [
+		ZFS_LINUX_COMPILE_IFELSE(
+		    [ZFS_LINUX_TEST_PROGRAM([[$1]], [[$2]])],
+		    [test -f build/conftest/conftest.ko], [$3], [$4])
+	])
```

Comment:
So, it's totally going over my head why this works.  It looks like we're doing the exact same `ZFS_LINUX_COMPILE_IFELSE()` lines if enable_linux_builtin is set or not, which seems to be functionally the same as the old code.  Can you explain what I'm missing here?

Reply:
@tonyhutter A different file is checked :smile:. 

`[test -f build/conftest/conftest.o]` vs `[test -f build/conftest/conftest.ko]` since when it's using the builtin case, it only creates the object file (`.o`), not the kernel module (`.ko`).

Reply:
That's right.

Reply:
Ah, need to get my eyes checked, thanks!